{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "928c2ec2",
   "metadata": {},
   "source": [
    "### Importing the libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f5a6501a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio.Seq import Seq\n",
    "from Bio.pairwise2 import align"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8794da5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "# List of file paths\n",
    "file_paths = [\"test2.extendedFrags.fastq\"]\n",
    "\n",
    "# List to store all reads from all files\n",
    "all_reads = []\n",
    "\n",
    "# Iterate over each file\n",
    "for file_path in file_paths:\n",
    "    # Use SeqIO to read the .fsa file\n",
    "    reads = list(SeqIO.parse(file_path, \"fastq\"))\n",
    "    \n",
    "    # Add the reads to the list of all reads\n",
    "    all_reads.extend(reads)\n",
    "\n",
    "# Now all_reads contains reads from all files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1eddf013",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "47186"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(all_reads)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0464ebba",
   "metadata": {},
   "source": [
    "### Maximum alignment score between all the sequences with different off-targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "18066797",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment Score (sequence 0 vs. sequence 1): 378.3\n",
      "Alignment Score (sequence 0 vs. sequence 2): 352.00000000000006\n",
      "Alignment Score (sequence 0 vs. sequence 3): 351.1\n",
      "Alignment Score (sequence 0 vs. sequence 4): 355.3\n",
      "Alignment Score (sequence 0 vs. sequence 5): 361.5\n",
      "Alignment Score (sequence 0 vs. sequence 6): 361.5\n",
      "Alignment Score (sequence 0 vs. sequence 7): 353.70000000000005\n",
      "Alignment Score (sequence 0 vs. sequence 8): 357.20000000000005\n",
      "Alignment Score (sequence 0 vs. sequence 9): 353.6\n",
      "Alignment Score (sequence 0 vs. sequence 10): 357.20000000000005\n",
      "Alignment Score (sequence 0 vs. sequence 11): 360.80000000000007\n",
      "Alignment Score (sequence 0 vs. sequence 12): 353.00000000000006\n",
      "Alignment Score (sequence 0 vs. sequence 13): 358.1\n",
      "Alignment Score (sequence 0 vs. sequence 14): 349.80000000000007\n",
      "Alignment Score (sequence 0 vs. sequence 15): 351.30000000000007\n",
      "Alignment Score (sequence 0 vs. sequence 16): 354.1\n",
      "Alignment Score (sequence 0 vs. sequence 17): 356.20000000000005\n",
      "Alignment Score (sequence 1 vs. sequence 2): 357.50000000000006\n",
      "Alignment Score (sequence 1 vs. sequence 3): 354.1\n",
      "Alignment Score (sequence 1 vs. sequence 4): 364.5\n",
      "Alignment Score (sequence 1 vs. sequence 5): 362.3\n",
      "Alignment Score (sequence 1 vs. sequence 6): 359.0\n",
      "Alignment Score (sequence 1 vs. sequence 7): 359.40000000000003\n",
      "Alignment Score (sequence 1 vs. sequence 8): 361.90000000000003\n",
      "Alignment Score (sequence 1 vs. sequence 9): 354.90000000000003\n",
      "Alignment Score (sequence 1 vs. sequence 10): 359.30000000000007\n",
      "Alignment Score (sequence 1 vs. sequence 11): 360.30000000000007\n",
      "Alignment Score (sequence 1 vs. sequence 12): 359.80000000000007\n",
      "Alignment Score (sequence 1 vs. sequence 13): 361.6\n",
      "Alignment Score (sequence 1 vs. sequence 14): 357.20000000000005\n",
      "Alignment Score (sequence 1 vs. sequence 15): 352.30000000000007\n",
      "Alignment Score (sequence 1 vs. sequence 16): 356.1\n",
      "Alignment Score (sequence 1 vs. sequence 17): 356.30000000000007\n",
      "Alignment Score (sequence 2 vs. sequence 3): 379.2\n",
      "Alignment Score (sequence 2 vs. sequence 4): 357.4000000000001\n",
      "Alignment Score (sequence 2 vs. sequence 5): 354.6\n",
      "Alignment Score (sequence 2 vs. sequence 6): 353.6\n",
      "Alignment Score (sequence 2 vs. sequence 7): 354.80000000000007\n",
      "Alignment Score (sequence 2 vs. sequence 8): 355.20000000000005\n",
      "Alignment Score (sequence 2 vs. sequence 9): 351.6\n",
      "Alignment Score (sequence 2 vs. sequence 10): 357.9000000000001\n",
      "Alignment Score (sequence 2 vs. sequence 11): 358.20000000000005\n",
      "Alignment Score (sequence 2 vs. sequence 12): 355.30000000000007\n",
      "Alignment Score (sequence 2 vs. sequence 13): 358.00000000000006\n",
      "Alignment Score (sequence 2 vs. sequence 14): 355.70000000000005\n",
      "Alignment Score (sequence 2 vs. sequence 15): 352.40000000000003\n",
      "Alignment Score (sequence 2 vs. sequence 16): 353.0\n",
      "Alignment Score (sequence 2 vs. sequence 17): 356.40000000000003\n",
      "Alignment Score (sequence 3 vs. sequence 4): 356.80000000000007\n",
      "Alignment Score (sequence 3 vs. sequence 5): 354.6\n",
      "Alignment Score (sequence 3 vs. sequence 6): 356.1\n",
      "Alignment Score (sequence 3 vs. sequence 7): 352.00000000000006\n",
      "Alignment Score (sequence 3 vs. sequence 8): 350.20000000000005\n",
      "Alignment Score (sequence 3 vs. sequence 9): 350.4000000000001\n",
      "Alignment Score (sequence 3 vs. sequence 10): 352.50000000000006\n",
      "Alignment Score (sequence 3 vs. sequence 11): 356.70000000000005\n",
      "Alignment Score (sequence 3 vs. sequence 12): 352.70000000000005\n",
      "Alignment Score (sequence 3 vs. sequence 13): 352.00000000000006\n",
      "Alignment Score (sequence 3 vs. sequence 14): 355.80000000000007\n",
      "Alignment Score (sequence 3 vs. sequence 15): 354.30000000000007\n",
      "Alignment Score (sequence 3 vs. sequence 16): 353.40000000000003\n",
      "Alignment Score (sequence 3 vs. sequence 17): 351.1\n",
      "Alignment Score (sequence 4 vs. sequence 5): 377.2\n",
      "Alignment Score (sequence 4 vs. sequence 6): 373.3\n",
      "Alignment Score (sequence 4 vs. sequence 7): 359.20000000000005\n",
      "Alignment Score (sequence 4 vs. sequence 8): 359.0\n",
      "Alignment Score (sequence 4 vs. sequence 9): 359.0\n",
      "Alignment Score (sequence 4 vs. sequence 10): 361.70000000000005\n",
      "Alignment Score (sequence 4 vs. sequence 11): 362.90000000000003\n",
      "Alignment Score (sequence 4 vs. sequence 12): 361.90000000000003\n",
      "Alignment Score (sequence 4 vs. sequence 13): 357.30000000000007\n",
      "Alignment Score (sequence 4 vs. sequence 14): 361.90000000000003\n",
      "Alignment Score (sequence 4 vs. sequence 15): 359.40000000000003\n",
      "Alignment Score (sequence 4 vs. sequence 16): 355.0\n",
      "Alignment Score (sequence 4 vs. sequence 17): 352.30000000000007\n",
      "Alignment Score (sequence 5 vs. sequence 6): 374.4\n",
      "Alignment Score (sequence 5 vs. sequence 7): 359.20000000000005\n",
      "Alignment Score (sequence 5 vs. sequence 8): 360.1\n",
      "Alignment Score (sequence 5 vs. sequence 9): 355.70000000000005\n",
      "Alignment Score (sequence 5 vs. sequence 10): 358.1\n",
      "Alignment Score (sequence 5 vs. sequence 11): 358.5\n",
      "Alignment Score (sequence 5 vs. sequence 12): 354.90000000000003\n",
      "Alignment Score (sequence 5 vs. sequence 13): 359.4\n",
      "Alignment Score (sequence 5 vs. sequence 14): 355.80000000000007\n",
      "Alignment Score (sequence 5 vs. sequence 15): 349.4000000000001\n",
      "Alignment Score (sequence 5 vs. sequence 16): 359.80000000000007\n",
      "Alignment Score (sequence 5 vs. sequence 17): 359.70000000000005\n",
      "Alignment Score (sequence 6 vs. sequence 7): 358.6\n",
      "Alignment Score (sequence 6 vs. sequence 8): 354.0\n",
      "Alignment Score (sequence 6 vs. sequence 9): 359.20000000000005\n",
      "Alignment Score (sequence 6 vs. sequence 10): 356.70000000000005\n",
      "Alignment Score (sequence 6 vs. sequence 11): 362.3\n",
      "Alignment Score (sequence 6 vs. sequence 12): 358.6\n",
      "Alignment Score (sequence 6 vs. sequence 13): 358.20000000000005\n",
      "Alignment Score (sequence 6 vs. sequence 14): 352.6\n",
      "Alignment Score (sequence 6 vs. sequence 15): 360.0\n",
      "Alignment Score (sequence 6 vs. sequence 16): 359.1\n",
      "Alignment Score (sequence 6 vs. sequence 17): 352.70000000000005\n",
      "Alignment Score (sequence 7 vs. sequence 8): 378.5\n",
      "Alignment Score (sequence 7 vs. sequence 9): 378.9\n",
      "Alignment Score (sequence 7 vs. sequence 10): 379.9\n",
      "Alignment Score (sequence 7 vs. sequence 11): 379.0\n",
      "Alignment Score (sequence 7 vs. sequence 12): 389.8\n",
      "Alignment Score (sequence 7 vs. sequence 13): 380.6\n",
      "Alignment Score (sequence 7 vs. sequence 14): 353.1\n",
      "Alignment Score (sequence 7 vs. sequence 15): 357.1\n",
      "Alignment Score (sequence 7 vs. sequence 16): 359.1\n",
      "Alignment Score (sequence 7 vs. sequence 17): 350.50000000000006\n",
      "Alignment Score (sequence 8 vs. sequence 9): 380.6\n",
      "Alignment Score (sequence 8 vs. sequence 10): 384.6\n",
      "Alignment Score (sequence 8 vs. sequence 11): 378.2\n",
      "Alignment Score (sequence 8 vs. sequence 12): 378.6\n",
      "Alignment Score (sequence 8 vs. sequence 13): 381.3\n",
      "Alignment Score (sequence 8 vs. sequence 14): 354.9000000000001\n",
      "Alignment Score (sequence 8 vs. sequence 15): 352.1\n",
      "Alignment Score (sequence 8 vs. sequence 16): 355.30000000000007\n",
      "Alignment Score (sequence 8 vs. sequence 17): 355.50000000000006\n",
      "Alignment Score (sequence 9 vs. sequence 10): 380.2\n",
      "Alignment Score (sequence 9 vs. sequence 11): 379.5\n",
      "Alignment Score (sequence 9 vs. sequence 12): 378.3\n",
      "Alignment Score (sequence 9 vs. sequence 13): 376.0\n",
      "Alignment Score (sequence 9 vs. sequence 14): 355.50000000000006\n",
      "Alignment Score (sequence 9 vs. sequence 15): 356.50000000000006\n",
      "Alignment Score (sequence 9 vs. sequence 16): 356.70000000000005\n",
      "Alignment Score (sequence 9 vs. sequence 17): 350.80000000000007\n",
      "Alignment Score (sequence 10 vs. sequence 11): 384.4\n",
      "Alignment Score (sequence 10 vs. sequence 12): 380.8\n",
      "Alignment Score (sequence 10 vs. sequence 13): 380.9\n",
      "Alignment Score (sequence 10 vs. sequence 14): 362.4000000000001\n",
      "Alignment Score (sequence 10 vs. sequence 15): 353.1\n",
      "Alignment Score (sequence 10 vs. sequence 16): 359.30000000000007\n",
      "Alignment Score (sequence 10 vs. sequence 17): 355.20000000000005\n",
      "Alignment Score (sequence 11 vs. sequence 12): 380.7\n",
      "Alignment Score (sequence 11 vs. sequence 13): 382.9\n",
      "Alignment Score (sequence 11 vs. sequence 14): 354.80000000000007\n",
      "Alignment Score (sequence 11 vs. sequence 15): 358.20000000000005\n",
      "Alignment Score (sequence 11 vs. sequence 16): 359.00000000000006\n",
      "Alignment Score (sequence 11 vs. sequence 17): 356.1\n",
      "Alignment Score (sequence 12 vs. sequence 13): 384.6\n",
      "Alignment Score (sequence 12 vs. sequence 14): 355.50000000000006\n",
      "Alignment Score (sequence 12 vs. sequence 15): 363.50000000000006\n",
      "Alignment Score (sequence 12 vs. sequence 16): 358.70000000000005\n",
      "Alignment Score (sequence 12 vs. sequence 17): 350.4000000000001\n",
      "Alignment Score (sequence 13 vs. sequence 14): 355.1\n",
      "Alignment Score (sequence 13 vs. sequence 15): 354.30000000000007\n",
      "Alignment Score (sequence 13 vs. sequence 16): 356.9000000000001\n",
      "Alignment Score (sequence 13 vs. sequence 17): 357.50000000000006\n",
      "Alignment Score (sequence 14 vs. sequence 15): 376.0\n",
      "Alignment Score (sequence 14 vs. sequence 16): 381.0\n",
      "Alignment Score (sequence 14 vs. sequence 17): 382.0\n",
      "Alignment Score (sequence 15 vs. sequence 16): 379.8\n",
      "Alignment Score (sequence 15 vs. sequence 17): 371.3\n",
      "Alignment Score (sequence 16 vs. sequence 17): 380.5\n",
      "Maximum Alignment Score: 389.8\n",
      "Aligned Sequences with Maximum Score: ('AGCCCTGGTGCCCTCACTATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTCCCATCACATCAACCGGGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', 'AGGCCCTTGCCTCAACCTATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTCCCATCACATCAACCGGGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA')\n"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "\n",
    "def perform_swalignment(seq1, seq2):\n",
    "    alignments = pairwise2.align.localms(seq1, seq2, 2, -1, -0.5, -0.1)\n",
    "    alignment = alignments[0]\n",
    "    alignment_score = alignment[2]\n",
    "    return alignment_score\n",
    "\n",
    "\n",
    "fullcheck_sequences =    ['TGGTCAGTGGGAAGGGACATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTTCGTGGCAATGCGCCACGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'AGCTGTCCACCTGTTGGAATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTTCGTGGCAATGCGCCACGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'CCCAGAAACCTCCAGTGGATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCGCCACCGGTTGATGTGATGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'CACACATACACTATGAAAATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCGCCACCGGTTGATGTGATGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'CACCTCCTACCTGTCGCCATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGTGGCGCATTGCCACGAAGCGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'CACCATGTGGGGAGTGTTATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGTGGCGCATTGCCACGAAGCGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'AGGAGTCAAAAGACTACAATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGTGGCGCATTGCCACGAAGCGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'AGCCCTGGTGCCCTCACTATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTCCCATCACATCAACCGGGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'TGGATTCCCCTGCTGGGTATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTCCCATCACATCAACCGGGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'TGGAGTCTATTTCCTCTCATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTCCCATCACATCAACCGGGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'TGGCCTCGACTCTGGTCTATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTCCCATCACATCAACCGGGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'GGGCAACTCAGCCTGGACATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTCCCATCACATCAACCGGGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'AGGCCCTTGCCTCAACCTATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTCCCATCACATCAACCGGGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'AGGGCCCTGTCAAGGTGGATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTCCCATCACATCAACCGGGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'CTCTGACAATCTGTCTTGATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGAGTCCGAGCAGAAGAAGAAGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'GAGAGCCACTACCCAACCATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGAGTCCGAGCAGAAGAAGAAGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'GAGGCAGCCTAGAGTCTTATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGAGTCCGAGCAGAAGAAGAAGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'CGGCGACAGATGTTGGGGATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGAGTCCGAGCAGAAGAAGAAGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA'\n",
    "                        ]\n",
    "\n",
    "\n",
    "max_alignment_score = float('-inf')\n",
    "max_alignment_pair = None\n",
    "\n",
    "for i in range(len(fullcheck_sequences)):\n",
    "    for j in range(i + 1, len(fullcheck_sequences)):\n",
    "        target_seq1 = fullcheck_sequences[i]\n",
    "        target_seq2 = fullcheck_sequences[j]\n",
    "\n",
    "        alignment_score = perform_swalignment(target_seq1, target_seq2)\n",
    "        print(f\"Alignment Score (sequence {i} vs. sequence {j}): {alignment_score}\")\n",
    "\n",
    "        if alignment_score > max_alignment_score:\n",
    "            max_alignment_score = alignment_score\n",
    "            max_alignment_pair = (target_seq1, target_seq2)\n",
    "\n",
    "print(f\"Maximum Alignment Score: {max_alignment_score}\")\n",
    "print(f\"Aligned Sequences with Maximum Score: {max_alignment_pair}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95bd15b5",
   "metadata": {},
   "source": [
    "### Importing alignment algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "411d72d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment Score: 29.299999999999997\n",
      "Aligned Sequence 1: -----------------CG--G------------TTGA---TG-TGAT-G-GG-AGGGAT-CTTATAG--A--\n",
      "Aligned Sequence 2: GACATCACCTCCCACAACGACGAAAACCTCCCCATTG-GCCTGCT--TCGTGGCA---ATGC-----GCCACC\n"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "def perform_swalignment(seq1, seq2):\n",
    "    alignments = pairwise2.align.localms(seq1, seq2, 2, -1, -0.5, -0.1)\n",
    "    \n",
    "    # Extract the first alignment (assuming you want the best one)\n",
    "    best_alignment = alignments[0]\n",
    "    \n",
    "    # Extract the aligned sequences and score\n",
    "    aligned_seq1 = best_alignment.seqA\n",
    "    aligned_seq2 = best_alignment.seqB\n",
    "    alignment_score = best_alignment.score\n",
    "    \n",
    "    return alignment_score, aligned_seq1, aligned_seq2\n",
    "\n",
    "# Example usage\n",
    "seq1 = Seq(\"CGGTTGATGTGATGGGAGGGATCTTATAGA\")\n",
    "seq2 = Seq(\"GACATCACCTCCCACAACGACGAAAACCTCCCCATTGGCCTGCTTCGTGGCAATGCGCCACC\")\n",
    "\n",
    "alignment_score, aligned_seq1, aligned_seq2 = perform_swalignment(seq1, seq2)\n",
    "\n",
    "print(f\"Alignment Score: {alignment_score}\")\n",
    "print(f\"Aligned Sequence 1: {aligned_seq1}\")\n",
    "print(f\"Aligned Sequence 2: {aligned_seq2}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f48ccafa",
   "metadata": {},
   "source": [
    "### Alignment Score of the sequence with itself to determine the threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ab9fcd74",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment Score: 398.0\n",
      "Aligned Sequence 1: GAGAGCCACTACCCAACCATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGAGTCCGAGCAGAAGAAGAAGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA\n",
      "Aligned Sequence 2: GAGAGCCACTACCCAACCATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGAGTCCGAGCAGAAGAAGAAGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA\n"
     ]
    }
   ],
   "source": [
    "seq_x = Seq('GAGAGCCACTACCCAACCATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGAGTCCGAGCAGAAGAAGAAGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA')\n",
    "seq_y = Seq('GAGAGCCACTACCCAACCATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGAGTCCGAGCAGAAGAAGAAGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA')\n",
    "alignment_score, aligned_seq1, aligned_seq2 = perform_swalignment(seq_x, seq_y)\n",
    "\n",
    "print(f\"Alignment Score: {alignment_score}\")\n",
    "print(f\"Aligned Sequence 1: {aligned_seq1}\")\n",
    "print(f\"Aligned Sequence 2: {aligned_seq2}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72a8eda8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Smallcut_all_sequences-Run10\n",
    "\n",
    "fullcheck_sequences =    ['TGGTCAGTGGGAAGGGACATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTTCGTGGCAATGCGCCACGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'AGCTGTCCACCTGTTGGAATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTTCGTGGCAATGCGCCACGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'CCCAGAAACCTCCAGTGGATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCGCCACCGGTTGATGTGATGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'CACACATACACTATGAAAATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCGCCACCGGTTGATGTGATGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'CACCTCCTACCTGTCGCCATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGTGGCGCATTGCCACGAAGCGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'CACCATGTGGGGAGTGTTATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGTGGCGCATTGCCACGAAGCGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'AGGAGTCAAAAGACTACAATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGTGGCGCATTGCCACGAAGCGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'AGCCCTGGTGCCCTCACTATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTCCCATCACATCAACCGGGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'TGGATTCCCCTGCTGGGTATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTCCCATCACATCAACCGGGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'TGGAGTCTATTTCCTCTCATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTCCCATCACATCAACCGGGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'TGGCCTCGACTCTGGTCTATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTCCCATCACATCAACCGGGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'GGGCAACTCAGCCTGGACATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTCCCATCACATCAACCGGGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'AGGCCCTTGCCTCAACCTATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTCCCATCACATCAACCGGGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'AGGGCCCTGTCAAGGTGGATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGCTCCCATCACATCAACCGGGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'CTCTGACAATCTGTCTTGATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGAGTCCGAGCAGAAGAAGAAGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'GAGAGCCACTACCCAACCATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGAGTCCGAGCAGAAGAAGAAGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'GAGGCAGCCTAGAGTCTTATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGAGTCCGAGCAGAAGAAGAAGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA', \n",
    "                         'CGGCGACAGATGTTGGGGATCTTATAGAGCCTGTGTGGACCCCCTATTTCTGATAATACGACTCACTATAGGAGTCCGAGCAGAAGAAGAAGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA'\n",
    "                        ]\n",
    "\n",
    "smallcut_identifier = ['TGGTCAGTGGGAAGGGA',\n",
    "                       'AGCTGTCCACCTGTTGG', \n",
    "                       'CCCAGAAACCTCCAGTG',\n",
    "                       'CACACATACACTATGAA', \n",
    "                       'CACCTCCTACCTGTCGC', \n",
    "                       'CACCATGTGGGGAGTGT', \n",
    "                       'AGGAGTCAAAAGACTAC', \n",
    "                       'AGCCCTGGTGCCCTCAC', \n",
    "                       'TGGATTCCCCTGCTGGG', \n",
    "                       'TGGAGTCTATTTCCTCT',\n",
    "                       'TGGCCTCGACTCTGGTC', \n",
    "                       'GGGCAACTCAGCCTGGA', \n",
    "                       'AGGCCCTTGCCTCAACC',\n",
    "                       'AGGGCCCTGTCAAGGTG', \n",
    "                       'CTCTGACAATCTGTCTT', \n",
    "                       'GAGAGCCACTACCCAAC', \n",
    "                       'GAGGCAGCCTAGAGTCT',\n",
    "                       'CGGCGACAGATGTTGGG']\n",
    "\n",
    "offtarget_rev = ['GTTGTGCATTGCCACAAAGC',\n",
    "                 'TCGTGTCAATTTGCCACGGG',\n",
    "                 'CCATGGGTTGATGTGATGAG',\n",
    "                 'CCACCAGTTGATCTGATGGG',\n",
    "                 'GCTCATTGCCAAGGAGCCGG',\n",
    "                 'TGGCATTGACACGAAGCTGA',\n",
    "                 'GCTTAGTGGCAATGTGCCAT',\n",
    "                 'CCCATCACATCACCCTGAGG',\n",
    "                 'CTGGTGGATGTGATGGGAGA',\n",
    "                 'CCTGTTGATGTGAGGGCAGC',\n",
    "                 'CCTGTGGATGTGATGGGATC',\n",
    "                 'CCGGCTGCTGTGATGGAAGC',\n",
    "                 'CCGGTAGATCTGATGGGGGC',\n",
    "                 'CAGGTTCATGTGAGGGGAGC',\n",
    "                 'TTAGAGCAGAAGAAGAAAGG',\n",
    "                 'TTCTTCTTCTGCTTAGACTC',\n",
    "                 'TTCTTCTCCTGCTAGGACTC',\n",
    "                 'TCTTTCTTCTGCTCGGCCTC']\n",
    "\n",
    "offtarget = ['GCTTTGTGGCAATGCACAAC',\n",
    "             'CCCGTGGCAAATTGACACGA',\n",
    "             'CTCATCACATCAACCCATGG',\n",
    "             'CCCATCAGATCAACTGGTGG',\n",
    "             'CCGGCTCCTTGGCAATGAGC',\n",
    "             'TCAGCTTCGTGTCAATGCCA',\n",
    "             'ATGGCACATTGCCACTAAGC',\n",
    "             'CCTCAGGGTGATGTGATGGG',\n",
    "             'TCTCCCATCACATCCACCAG',\n",
    "             'GCTGCCCTCACATCAACAGG',\n",
    "             'GATCCCATCACATCCACAGG',\n",
    "             'GCTTCCATCACAGCAGCCGG',\n",
    "             'GCCCCCATCAGATCTACCGG',\n",
    "             'GCTCCCCTCACATGAACCTG',\n",
    "             'CCTTTCTTCTTCTGCTCTAA',\n",
    "             'GAGTCTAAGCAGAAGAAGAA',\n",
    "             'GAGTCCTAGCAGGAGAAGAA',\n",
    "             'GAGGCCGAGCAGAAGAAAGA']\n",
    "\n",
    "\n",
    "primer = 'GACATCACCTCCCACAACGACGAAAA'\n",
    "\n",
    "\n",
    "# Assuming you have the perform_swalignment function already defined\n",
    "\n",
    "check = 'GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAACGAGGCGAGTTTACGGGTTGTTA'\n",
    "\n",
    "smallcut_dict = {}\n",
    "smallcut_dict1 = {}\n",
    "for j in range(18):\n",
    "    smallcut = []\n",
    "    smallcut1 = []\n",
    "\n",
    "    for record in all_reads:\n",
    "        s1 = str(record.seq)\n",
    "        #print(s1)\n",
    "        s2 = str(Seq(s1).reverse_complement())\n",
    "        L = len(s1)\n",
    "        #if (L >= 100 and L < 250):\n",
    "        if L >= 100:\n",
    "            sc, _, _ = perform_swalignment(Seq(s1[:]), Seq(check))\n",
    "            #print('sc',sc)\n",
    "            sr, _, _ = perform_swalignment(Seq(s2[:]), Seq(check))\n",
    "            #print('sr',sr)\n",
    "            if sr >= 200 and sr > sc:\n",
    "                s1 = s2\n",
    "                sc = sr\n",
    "            elif sc >= 200 :\n",
    "                sc2, _, _ = perform_swalignment(Seq(s1[:]), Seq(fullcheck_sequences[j]))\n",
    "                print('sc2',sc2)\n",
    "                #ON1 = s1.find(fullcheck_sequences[j]) \n",
    "                if sc2 >= 390:\n",
    "\n",
    "                    smallcut1.append(s1)\n",
    "                    smallcut.append(record)\n",
    "                    print(f' found small cut {j}')\n",
    "\n",
    "     \n",
    "    smallcut_dict1[j] = smallcut1\n",
    "    smallcut_dict[j] = smallcut\n",
    "    \n",
    "# Save bigcut lists for each check sequence\n",
    "\n",
    "for j, smallcut in smallcut_dict.items():\n",
    "    filename = str(j) + 'Run1_smallcut.fasta'\n",
    "    SeqIO.write(smallcut, filename, 'fasta')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58dae325",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the lists for each check sequence\n",
    "\n",
    "for j, smallcut in smallcut_dict.items():\n",
    "    filename = str(j) + 'Run1_smallcut.fasta'\n",
    "    SeqIO.write(smallcut, filename, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02f3438c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Print the number of elements for each set\n",
    "for i in range(18):\n",
    "    print(f\"Elements for Sequence {i+1}: {len(smallcut_dict1[i])}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd2b2179",
   "metadata": {},
   "outputs": [],
   "source": [
    "len('GACATCACCTCCCACAACGA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0eb7a93e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Finding Total bigcut and smallcut together excluding uncut\n",
    "def find_subsets(string):\n",
    "    subsets = []\n",
    "    for i in range(len(string)):\n",
    "        for j in range(i + 1, len(string) + 1):\n",
    "            if j - i >= 20:  # Check if the length of the substring is at least 5\n",
    "                subsets.append(string[i:j])\n",
    "    return subsets\n",
    "\n",
    "# Example usage:\n",
    "subsets = find_subsets('GACATCACCTCCCACAACGA')\n",
    "subsets1 = find_subsets('TCGTTGTGGGAGGTGATGTC')\n",
    "# Find elements of the input sequences that are not substrings of specified length for each set\n",
    "elements_lists = [[] for _ in range(18)]\n",
    "for i in range(18):\n",
    "    for seq in smallcut_dict1[i]:   \n",
    "        if len(seq) >= 100:\n",
    "            found = False\n",
    "            for sub in subsets:\n",
    "                if sub in seq:\n",
    "                    found = True\n",
    "                    break\n",
    "            if not found:\n",
    "                elements_lists[i].append(seq)\n",
    "                \n",
    "elements_lists1 = [[] for _ in range(18)]\n",
    "for i in range(18):\n",
    "    for seq in elements_lists[i]:   \n",
    "        if len(seq) >= 100:\n",
    "            found = False\n",
    "            for sub in subsets1:\n",
    "                if sub in seq:\n",
    "                    found = True\n",
    "                    break\n",
    "            if not found:\n",
    "                elements_lists1[i].append(seq)\n",
    "\n",
    "# Print the number of elements for each set\n",
    "for i in range(18):\n",
    "    print(f\"Elements of the input sequences for set {i+1}: {len(elements_lists1[i])}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d9cd6c74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len('GCTTTGTGGCAATGCACAAC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bc8c50e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Finding number of small cut \n",
    "\n",
    "\n",
    "offtarget = ['GCTTTGTGGCAATGCACAAC',\n",
    "             'CCCGTGGCAAATTGACACGA',\n",
    "             'CTCATCACATCAACCCATGG',\n",
    "             'CCCATCAGATCAACTGGTGG',\n",
    "             'CCGGCTCCTTGGCAATGAGC',\n",
    "             'TCAGCTTCGTGTCAATGCCA',\n",
    "             'ATGGCACATTGCCACTAAGC',\n",
    "             'CCTCAGGGTGATGTGATGGG',\n",
    "             'TCTCCCATCACATCCACCAG',\n",
    "             'GCTGCCCTCACATCAACAGG',\n",
    "             'GATCCCATCACATCCACAGG',\n",
    "             'GCTTCCATCACAGCAGCCGG',\n",
    "             'GCCCCCATCAGATCTACCGG',\n",
    "             'GCTCCCCTCACATGAACCTG',\n",
    "             'CCTTTCTTCTTCTGCTCTAA',\n",
    "             'GAGTCTAAGCAGAAGAAGAA',\n",
    "             'GAGTCCTAGCAGGAGAAGAA',\n",
    "             'GAGGCCGAGCAGAAGAAAGA']\n",
    "\n",
    "\n",
    "def find_subsets(string):\n",
    "    subsets = []\n",
    "    for i in range(len(string)):\n",
    "        for j in range(i + 1, len(string) + 1):\n",
    "            if j - i >= 15:  # Check if the length of the substring is at least 5\n",
    "                subsets.append(string[i:j])\n",
    "    return subsets\n",
    "\n",
    "for i in range(18):\n",
    "    # Example usage:\n",
    "    subsets2 = find_subsets(offtarget[i])\n",
    "    # Find elements of the input sequences that are not substrings of specified length for each set\n",
    "    elements_lists2 = [[] for _ in range(18)]\n",
    "    for i in range(18):\n",
    "        for seq in elements_lists1[i]:   \n",
    "            if len(seq) >= 100:\n",
    "                found = False\n",
    "                for sub2 in subsets2:\n",
    "                    if sub2 in seq:\n",
    "                        found = True\n",
    "                        break\n",
    "                if not found:\n",
    "                    elements_lists2[i].append(seq)\n",
    "\n",
    "# Print the number of elements for each set\n",
    "for i in range(18):\n",
    "    print(f\"Elements of the input sequences for set {i+1}: {len(elements_lists2[i])}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c530cc38",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cdac24e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
